src/borg/archive.py

```diff
@@ -1052,6 +1052,19 @@ def verify_data(self):
                 self.error_found = True
                 errors += 1
                 logger.error('chunk %s, integrity error: %s', bin_to_hex(chunk_id), integrity_error)
+                if self.repair:
+                    # if we kill the defect chunk here, subsequent actions within this "borg check"
+                    # run will find missing chunks and replace them with all-zero replacement
+                    # chunks and flag the files as "repaired".
+                    # if another backup is done later and the missing chunks get backupped again,
+                    # a "borg check" afterwards can heal all files where this chunk was missing.
+                    logger.warning('chunk %s is being deleted, so affected files can get repaired now '
+                                   'and maybe healed later.', bin_to_hex(chunk_id))
+                    del self.chunks[chunk_id]
```

Comment:
this is unsafe [modify-collection-that-is-iterated-over], here specifically because self.chunks is the usual hashindex and the iterator is based on the internal modulus index. If the hashtable is resized we may skip stuff.

Solution: Add list of deleted chunk ids, remove them from self.chunks after the loop is done.